fs/base.py

```diff
@@ -599,14 +599,16 @@ def getsize(self, path):
         size = self.getdetails(path).size
         return size
 
-    def getsyspath(self, path):
+    def getsyspath(self, path, as_bytes=False):
```

Comment:
I'm not overly keen on overloading `getsyspath` like this.

How about leaving `getsyspath` as returning unicode, and add a `getnativepath` which may return any format (presumably bytes for *nix).

Reply:
@coveralls your call. I had started doing so but then I found it heavy: to have another method, another error, etc. And since your docstring for `getsyspath` states:

```
        A system path is one recognized by the OS, that may be used
        outside of PyFilesystem (in an application or a shell for
        example). This method will get the corresponding system path
        that would be referenced by ``path``.
```

... I find it really hard to have another method with different docstring... since on *nix the real system path is bytes.

Reply:
So what would happen if you were to get a syspath with `as_bytes=False` where the path couldn't be decoded as unicode?

Reply:
`as_bytes=False` is the default so you still always get `unicode` out. With `as_bytes=True` you will always get `bytes` out using the filesystem encoding.  There is no case I can think of where there would be nay possible issue (unless if you have no filesystem encoding at all.... but that is the source of so many possible errors and is a gross misconfiguration of the OS that's it is not worth catering to)

Reply:
but again, your call if you think that would break some API. I think not since this is only additive with a default.

Reply:
Will need to give this one some thought!

Reply:
Note also that strictly speaking the conversion to bytes on *nix is only required on Python2. Python3 handles this fine internally (and the os.path and related stdlib have been sprinkled with plenty of os.fsencode/fsdecode as needed) .... So I could make the test for both `*nix and py2` ... now the need to get bytes is still valid on Py2 and Py3 as a unicode fsdecoded path such as in #120 cannot be used in a *nix shell outside of Python unless it is fsencoded first.

Reply:
The latest commit only deals with bytes internally on Python 2 and *nix. Everywhere else, this is unicode throughout. We still need of course the ability to fsencode some unicode, but frankly this could be left out entirely of the `getsyspath` method. Instead we could update the doc and stating that `getsyspath` always return Unicode and when and why this could be a problem, and how to `fsencode` the received unicode for use outside if needed.

This could be a happy middle ground.

Reply:
@pombredanne Sounds good. But could you drop the `as_bytes` then? Calling this method with the new parameter will break with most of the other filesystems and all of the external filesystems.

I'm still favouring the `getnativepath` which may return bytes or unicode, but leaving `getsyspath` as returning (surrogate encoded) unicode. The default `getnativepath` could just return the same value as `getsyspath`

Reply:
> I'm still favouring the getnativepath which may return bytes or unicode, but leaving getsyspath as returning (surrogate encoded) unicode. The default getnativepath could just return the same value as getsyspath

Do you want this at all? I thing this is no longer needed as the doc explains that getsystempath is always unicode and that some use cases may need fsencoding to be usable